pkg/kubernetes/artefact_downloader.go

```diff
@@ -0,0 +1,128 @@
+package kubernetes
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"path/filepath"
+	"strings"
+
+	"github.com/suse-edge/edge-image-builder/pkg/http"
+	"github.com/suse-edge/edge-image-builder/pkg/image"
+	"github.com/suse-edge/edge-image-builder/pkg/log"
+	"golang.org/x/sync/errgroup"
+)
+
+const (
+	kubernetesDir = "kubernetes"
+
+	rke2ReleaseURL = "https://github.com/rancher/rke2/releases/download/%s/%s"
```

Comment:
Similar to my other comment... do we want to consider allowing the user to side-load the rke2/k3s binary rather than pulling the release during build?

Reply:
Sure but not right away. The goal is to get RKE2 up and running as soon as possible and cover the HA approach. There's no time to include both approaches right now and if I had to pick one, it'd be the automated way introduced in this PR. What are your thoughts?

Reply:
There's a third option: we skip the automated download entirely and instead the instructions are only for side-loading. That's somewhat consistent with how we handle the base image download, so there's a precedent for telling users to download things on their own. That simplifies what EIB has to do, makes testing easier, and removes the concerns about downloading timeouts. It is also a partial answer to the caching issue; the user downloads it once on their own instead of EIB downloading it on every build (at least every build with k8s configured).

@rdoxenham @atanasdinov  Thoughts?

Reply:
I'm not a fan of it. The issue is that the value EIB brings in this way is much lower - the user would have to get familiar with RKE2 installation docs, airgap methods, CNI artefacts matrix, etc. I much prefer EIB to do the heavy lifting at the cost of no caching until later point.

This being consistent with the other flows doesn't make it better either. @rdoxenham already suggested using the downloading logic to cover the Elemental RPMs too so we're actively striving to get away from side loading unless there's a clear benefit.

I'm also not concerned with downloading timeouts - clients would have to download the artefacts regardless so there's no workaround. After all, we're talking about downloads during build time, not during combustion time.

Last but not least, this doesn't make testing easier. Regardless of whether we document the flow or automate it, it'd still have to run through pipelines which will download the necessary files. Am I missing something?

Reply:
I'm not sure I understand your first point. The user would only need to know which files to download, it shouldn't require them to know specifics of airgap methods and so forth. That said, it does shift the burden over to docs for us to clearly explain that to them, which isn't great.

My timeout concern is perhaps unfounded. My biggest worry there is EIB perpetually hanging at a download step, but admittedly the likelihood is low.

My argument about making testing easier was twofold:
- there's less moving parts inside of EIB to worry about
- it'll be faster

I admit that last one is a crap reason to make a decision choice from. But I keep thinking back to the user experience. If they don't nail the image definition and configuration on the first try, they are going to need to retrigger builds. Perhaps frequently depending on how many iterations they are doing. My worry is that will sour them on EIB and push them away from using it if it's too time consuming and cumbersome to work with. It's why I've been talking about caching from the start instead of having that as an evolution.

But that's a gut feeling that I cannot back up in any way. Maybe users will understand that the act of building an all-in-one, air-gapped image _should_ take a while and have more patience for it. Hell, if we go with the third option, it could go the other way and have people turned off because they have to do so many manual prep steps.

All that said, I'm fine dropping my proposed third option and sticking with the download and then adding caching and side-loading later.

Reply:
> The user would only need to know which files to download, it shouldn't require them to know specifics of airgap methods and so forth.

They will have to know which tarballs to download i.e. all the logic of the `gatherArtefacts()` function. Depending on the CNI specifics and instruction set, those vary greatly. You'd then have to enforce the users not to rename the files in any way otherwise you'd be getting all sorts of different errors. Such validation itself in EIB would be a pain to introduce and maintain and we'd be running into more issues trying to understand why specific builds fail when the user doesn't follow the documentation properly.

> My biggest worry there is EIB perpetually hanging at a download step, but admittedly the likelihood is low.

This is not the case. The implementation I've proposed starts separate goroutines for each individual download. The second one of those fails due to any kind of error (network connectivity, disk space, etc.) all the go routines exit immediately and the build process stops.

As mentioned below, in order to make this process not a "freeze" experience, we should introduce the progress bar functionality where it explicitly showcases what step we're on and what exactly it is doing.

> If they don't nail the image definition and configuration on the first try, they are going to need to retrigger builds. Perhaps frequently depending on how many iterations they are doing.

This is an absolutely valid point. My look on it is different though. By making the builds faster to run and leveraging more manual input, the chance for user error is significantly greater. In the example above, if a user downloads an artefact and renames it, their build time will be extremely fast on running EIB. However, they won't be able to figure out what the error during deployment is unless they contact us to troubleshoot it. And now those few minutes we "saved" will turn into hours of back and forth discussions, debugging and figuring out how to reproduce a problem we'd never otherwise run into.

---

I appreciate I might be wrong here but I'm always against introducing more manual steps since the validation, testing and documentation on those is _usually_ much more a pain to deal with.

I'd love to hear @rdoxenham's opinion on all of this though.

Reply:
I'd like to get to the point where the user *CAN* side-load the artefacts if they would like to, e.g. because they've already got them and they don't want the tool to download them (although they would risk a version mismatch if they don't get the exact versions) to save bandwidth, etc. but for 3.0 we should be doing it automatically and driving towards guaranteed correct bits being in place. Timeouts are somewhat useful, but referring to my previous point, we have to recognise that some of our customers might not have a lot of bandwidth, so we should definitely consider caching between runs if we can verify the checksum in a cached directory. I think what's going to be absolutely key is that we're providing output on how the download is going (as you've said), and not just in the log files, but directly so the user knows what's happening whilst potentially large file downloads are taking place.